Store less raw pointers in containers in Source/rendering/

cdumez opened the pull request:
#### 8650e88882f120e9a711764b2bb85ba75356b0bc
<pre>
Store less raw pointers in containers in Source/rendering/
<a href="https://bugs.webkit.org/show_bug.cgi?id=261545">https://bugs.webkit.org/show_bug.cgi?id=261545</a>

Reviewed by Darin Adler.

* Source/WTF/wtf/WeakHashMap.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::layoutOverflowRectContainsAllDescendants):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::adjustVerticalPageScrollStepForFixedContent):
* Source/WebCore/rendering/InlineBoxPainter.cpp:
(WebCore::InlineBoxPainter::paint):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::insertIntoTrackedRendererMaps):
(WebCore::removeFromTrackedRendererMaps):
(WebCore::PositionedDescendantsMap::addDescendant):
(WebCore::PositionedDescendantsMap::removeDescendant):
(WebCore::PositionedDescendantsMap::removeContainingBlock):
(WebCore::PositionedDescendantsMap::positionedRenderers const):
(WebCore::removeBlockFromPercentageDescendantAndContainerMaps):
(WebCore::RenderBlock::~RenderBlock):
(WebCore::RenderBlock::blockWillBeDestroyed):
(WebCore::RenderBlock::hasRareData const):
(WebCore::getBlockRareData):
(WebCore::ensureBlockRareData):
(WebCore::RenderBlock::addOverflowFromPositionedObjects):
(WebCore::RenderBlock::dirtyForLayoutFromPercentageHeightDescendants):
(WebCore::RenderBlock::layoutPositionedObjects):
(WebCore::RenderBlock::markPositionedObjectsForLayout):
(WebCore::RenderBlock::paintObject):
(WebCore::RenderBlock::addContinuationWithOutline):
(WebCore::RenderBlock::paintsContinuationOutline):
(WebCore::RenderBlock::paintContinuationOutlines):
(WebCore::clipOutPositionedObjects):
(WebCore::RenderBlock::removePositionedObjects):
(WebCore::RenderBlock::percentHeightDescendants const):
(WebCore::RenderBlock::hasPercentHeightDescendant):
(WebCore::RenderBlock::checkPositionedObjectsNeedLayout):
* Source/WebCore/rendering/RenderBlock.h:
(WebCore::RenderBlock::hasPositionedObjects const):
(WebCore::RenderBlock::hasPercentHeightDescendants const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::rebuildFloatingObjectSetFromIntrudingFloats):
(WebCore::RenderBlockFlow::simplifiedNormalFlowLayout):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::childHasPercentHeightDescendants const):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::willBeDestroyed):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::relayoutCellIfFlexed):
(WebCore::RenderTableSection::paintObject):
* Source/WebCore/rendering/SelectionRangeData.cpp:
(WebCore::SelectionRangeData::repaint const):

Canonical link: <a href="https://commits.webkit.org/268000@main">https://commits.webkit.org/268000@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/109001e86f5289e7e816f9b9e099ba7122edd08b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/18285 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/18622 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/19201 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/20127 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/17116 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/18482 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/21920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/18775 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19051 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/18509 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/18718 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15927 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/21010 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15941 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/16681 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/23174 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/15845 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16960 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/16851 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21064 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/17572 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/17416 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/14774 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/21644 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/16511 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/5303 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4368 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/20875 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/22878 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/17259 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5150 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -107,32 +109,32 @@ static void insertIntoTrackedRendererMaps(const RenderBlock& container, RenderBo
     
-    HashSet<const RenderBlock*> containerSet = percentHeightContainerMap->take(&descendant);
-    for (auto* container : containerSet) {
+    WeakHashSet<const RenderBlock> containerSet = percentHeightContainerMap->take(descendant);
```
I suggest either `auto` or putting the `get` into the `for` statement instead of a local.

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -154,82 +156,82 @@ class PositionedDescendantsMap {
                 for (auto it = descendants->begin(); it != descendants->end(); ++it) {
-                    if ((*it)->isFixedPositioned()) {
-                        isNewEntry = descendants->insertBefore(it, &positionedDescendant).isNewEntry;
+                    if ((*it).isFixedPositioned()) {
```
Consider `it->` here?

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -241,7 +243,7 @@ static PositionedDescendantsMap& positionedDescendantsMap()
 }
 
-typedef HashMap<RenderBlock*, std::unique_ptr<ListHashSet<RenderInline*>>> ContinuationOutlineTableMap;
+using ContinuationOutlineTableMap = WeakHashMap<RenderBlock, std::unique_ptr<WeakListHashSet<RenderInline>>>;
```
I believe that `WeakListHashSet` is a single possibly null pointer to a hash table, it is wasteful to store it in turn in a `unique_ptr`. We should consider fixing this mistake at some point.

cdumez replied:
`WeakListHashSet` is 3 pointers (hashTable, m_head, m_tail) + an unsigned (m_operationCountSinceLastCleanup). It is not quite as trivial as a single pointer.

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -94,8 +96,8 @@ struct SameSizeAsRenderBlock : public RenderBox {
 
-typedef HashMap<const RenderBlock*, std::unique_ptr<TrackedRendererListHashSet>> TrackedDescendantsMap;
-typedef HashMap<const RenderBox*, HashSet<const RenderBlock*>> TrackedContainerMap;
+using TrackedDescendantsMap = WeakHashMap<const RenderBlock, std::unique_ptr<TrackedRendererListHashSet>>;
```
I believe that `TrackrdRendererListHashSet` is a single possibly null pointer to a hash table, and it is wasteful to store it in turn in a `unique_ptr`. We should consider fixing this mistake at some point.

cdumez replied:
It is a pointer to the HashTable and 2 unsigned data members.

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -154,82 +156,82 @@ class PositionedDescendantsMap {
 private:
-    using DescendantsMap = HashMap<const RenderBlock*, std::unique_ptr<TrackedRendererListHashSet>>;
-    using ContainerMap = HashMap<const RenderBox*, const RenderBlock*>;
+    using DescendantsMap = WeakHashMap<const RenderBlock, std::unique_ptr<TrackedRendererListHashSet>>;
```
I believe that `TrackrdRendererListHashSet` is a single possibly null pointer to a hash table, and it is wasteful to store it in turn in a `unique_ptr`. We should consider fixing this mistake at some point.

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -744,11 +746,11 @@ void RenderBlock::addOverflowFromPositionedObjects()
 
     for (auto it = positionedDescendants->begin(), end = positionedDescendants->end(); it != end; ++it) {
-        RenderBox* positionedObject = *it;
+        RenderBox& positionedObject = *it;
```
This is an old fashioned way of writing:
```
for (auto& positionedObject : *positionedDescendants)
```

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -811,12 +813,12 @@ void RenderBlock::dirtyForLayoutFromPercentageHeightDescendants()
         return;
 
-    TrackedRendererListHashSet* descendants = percentHeightDescendantsMap->get(this);
+    TrackedRendererListHashSet* descendants = percentHeightDescendantsMap->get(*this);
```
`auto`

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -811,12 +813,12 @@ void RenderBlock::dirtyForLayoutFromPercentageHeightDescendants()
 
     for (auto it = descendants->begin(), end = descendants->end(); it != end; ++it) {
-        RenderElement* renderer = *it;
+        RenderElement* renderer = &*it;
```
More like this?
```
for (auto& renderer : *descendants)
```

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -1347,29 +1347,29 @@ static ContinuationOutlineTableMap* continuationOutlineTable()
     
     ContinuationOutlineTableMap* table = continuationOutlineTable();
-    ListHashSet<RenderInline*>* continuations = table->get(this);
+    WeakListHashSet<RenderInline>* continuations = table->get(*this);
```
`auto`

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -1347,29 +1347,29 @@ static ContinuationOutlineTableMap* continuationOutlineTable()
-        
-    ListHashSet<RenderInline*>* continuations = table->get(this);
+
+    WeakListHashSet<RenderInline>* continuations = table->get(*this);
```
`auto`

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -1347,29 +1347,29 @@ static ContinuationOutlineTableMap* continuationOutlineTable()
 {
     ContinuationOutlineTableMap* table = continuationOutlineTable();
-    if (table->isEmpty())
+    if (table->isEmptyIgnoringNullReferences())
```
This check is wasteful and should be removed.

darinadler commented on Source/WebCore/rendering/RenderBlock.cpp:
```diff
@@ -1347,29 +1347,29 @@ static ContinuationOutlineTableMap* continuationOutlineTable()
-bool RenderBlock::paintsContinuationOutline(RenderInline* flow)
+bool RenderBlock::paintsContinuationOutline(RenderInline& flow)
 {
     ContinuationOutlineTableMap* table = continuationOutlineTable();
```
`auto`

darinadler commented on Source/WebCore/rendering/RenderBlock.h:
```diff
@@ -27,7 +27,7 @@
 #include "TextRun.h"
 #include <memory>
-#include <wtf/ListHashSet.h>
+#include <wtf/WeakListHashSet.h>
```
It's unfortunate for build times that this and other big headers have to be included in this header. Consider refactoring so it's not at some point? I think the issue is that there are many inline functions that access these sets. But there are also many that do not so some day we can improve this.

darinadler commented on Source/WebCore/rendering/RenderBlock.h:
```diff
@@ -39,7 +39,7 @@ class RenderText;
 struct PaintInfo;
 
-typedef ListHashSet<RenderBox*> TrackedRendererListHashSet;
+typedef WeakListHashSet<RenderBox> TrackedRendererListHashSet;
```
`using`

Also not a great type name. Not sure why a set type should be named a "tracked list" even if the set is used to "track a list of things".

darinadler commented on Source/WebCore/rendering/RenderBlock.h:
```diff
@@ -93,7 +93,7 @@ class RenderBlock : public RenderBox {
     bool hasPercentHeightDescendants() const
     {
         TrackedRendererListHashSet* objects = percentHeightDescendants();
```
`auto`

darinadler commented on Source/WebCore/rendering/RenderBlock.h:
```diff
@@ -84,7 +84,7 @@ class RenderBlock : public RenderBox {
     bool hasPositionedObjects() const
     {
         TrackedRendererListHashSet* objects = positionedObjects();
```
`auto`

cdumez committed:
Store less raw pointers in containers in Source/rendering/
https://bugs.webkit.org/show_bug.cgi?id=261545

Reviewed by Darin Adler.

* Source/WTF/wtf/WeakHashMap.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::layoutOverflowRectContainsAllDescendants):
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::adjustVerticalPageScrollStepForFixedContent):
* Source/WebCore/rendering/InlineBoxPainter.cpp:
(WebCore::InlineBoxPainter::paint):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::insertIntoTrackedRendererMaps):
(WebCore::removeFromTrackedRendererMaps):
(WebCore::PositionedDescendantsMap::addDescendant):
(WebCore::PositionedDescendantsMap::removeDescendant):
(WebCore::PositionedDescendantsMap::removeContainingBlock):
(WebCore::PositionedDescendantsMap::positionedRenderers const):
(WebCore::removeBlockFromPercentageDescendantAndContainerMaps):
(WebCore::RenderBlock::~RenderBlock):
(WebCore::RenderBlock::blockWillBeDestroyed):
(WebCore::RenderBlock::hasRareData const):
(WebCore::getBlockRareData):
(WebCore::ensureBlockRareData):
(WebCore::RenderBlock::addOverflowFromPositionedObjects):
(WebCore::RenderBlock::dirtyForLayoutFromPercentageHeightDescendants):
(WebCore::RenderBlock::layoutPositionedObjects):
(WebCore::RenderBlock::markPositionedObjectsForLayout):
(WebCore::RenderBlock::paintObject):
(WebCore::RenderBlock::addContinuationWithOutline):
(WebCore::RenderBlock::paintsContinuationOutline):
(WebCore::RenderBlock::paintContinuationOutlines):
(WebCore::clipOutPositionedObjects):
(WebCore::RenderBlock::removePositionedObjects):
(WebCore::RenderBlock::percentHeightDescendants const):
(WebCore::RenderBlock::hasPercentHeightDescendant):
(WebCore::RenderBlock::checkPositionedObjectsNeedLayout):
* Source/WebCore/rendering/RenderBlock.h:
(WebCore::RenderBlock::hasPositionedObjects const):
(WebCore::RenderBlock::hasPercentHeightDescendants const):
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::rebuildFloatingObjectSetFromIntrudingFloats):
(WebCore::RenderBlockFlow::simplifiedNormalFlowLayout):
* Source/WebCore/rendering/RenderFlexibleBox.cpp:
(WebCore::RenderFlexibleBox::childHasPercentHeightDescendants const):
* Source/WebCore/rendering/RenderInline.cpp:
(WebCore::RenderInline::willBeDestroyed):
* Source/WebCore/rendering/RenderTableSection.cpp:
(WebCore::RenderTableSection::relayoutCellIfFlexed):
(WebCore::RenderTableSection::paintObject):
* Source/WebCore/rendering/SelectionRangeData.cpp:
(WebCore::SelectionRangeData::repaint const):

Canonical link: https://commits.webkit.org/268000@main
